### PR TITLE
Misc changes to speedup gatsbyjs.org

### PR DESCRIPTION
**Pull request by KyleAMathews**
There's a lot here:

- lazy load twitter scripts — only load them if we see an embedded tweet. This saves loading a few dozen kbs of JS on page load
- fix bug in gatsby-plugin-offline where we were eval/parsing all prefetched resources for other pages before the service worker had been installed. This caused a lot of extra CPU load on the main thread. Fixed it so the plugin only fetches the resources and doesn't eval them.
- completely moved away from the v1 prefetching queue algorithm. We weren't using it anyways and we moved to a far better way of prefetching where now we use `<link prefetch>` for lazy loading resources for linked pages. This is the widely supported browser feature for this sort of thing. No idea why I didn't use this last year 🙈In tests, just starting prefetching when links come into view works quite well. It also gets around the problem that the old prefetching queue was trying to solve, that is, not blocking loading high priority downloads for the current page, as browsers set `<link prefetch>` downloads at the lowest priority and pull network resources from them as needed.
- lazy load algolia JS when the user interacts with the search form

### CODE REVIEW

File: packages/gatsby/cache-dir/loader.js [hunk 1]
```diff
@@ -241,41 +228,31 @@ const queue = {
...
-
-    // Add resources to queue.
-    enqueueResource(page.jsonName)
-    enqueueResource(page.componentChunkName)
-
-    // Sort resources by resourcesCount.
-    resourcesArray.sort(sortResourcesByCount)
-
+    // Prefetch resources.
+    prefetchResource(page.jsonName)
```
Comment (KyleAMathews) [inline packages/gatsby/cache-dir/loader.js]: Just prefetch immediately

File: packages/gatsby/cache-dir/prefetch.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  parentElement.appendChild(link)
+}
+const importPrefetchStrategy = function(url) {
+  return new Promise(function(resolve_1, reject_1) {
+    require([url], resolve_1, reject_1)
+  })
+}
+const supportedPrefetchStrategy = support("prefetch")
+  ? linkPrefetchStrategy
+  : importPrefetchStrategy
```
Comment (KyleAMathews) [inline packages/gatsby/cache-dir/prefetch.js]: only Safari doesn't support prefetch (wonder why)

File: packages/gatsby/cache-dir/static-entry.js [hunk 3]
```diff
@@ -329,6 +333,21 @@ export default (pagePath, callback) => {
     />
   )
 
+  // Add chunk mapping metadata
+  const scriptChunkMapping = `/*<![CDATA[*/window.___chunkMapping=${JSON.stringify(
```
Comment (KyleAMathews) [inline packages/gatsby/cache-dir/static-entry.js]: add chunk mappings to html (mapping from the chunk names to the actual path of the js file) for component chunks.

File: www/src/components/search-form.js [hunk 4]
```diff
@@ -292,6 +310,8 @@ class SearchForm extends Component {
           marginBottom: 0,
         }}
         className="searchWrap"
+        onMouseOver={() => this.loadAlgoliaJS()}
```
Comment (KyleAMathews) [inline www/src/components/search-form.js]: lazy load algolia js when the user mouse over or clicks the form

File: packages/gatsby-plugin-twitter/src/gatsby-browser.js [hunk 5]
```diff
@@ -1,9 +1,46 @@
...
-    window.twttr.widgets &&
-    typeof window.twttr.widgets.load === `function`
-  ) {
-    window.twttr.widgets.load()
+  // If there's an embedded tweet, lazy-load the twitter script (if it hasn't
+  // already been loaded), and then run the twitter load function.
+  if (document.querySelector(`.twitter-tweet`) !== null) {
+    if (!injectedTwitterScript) {
+      injectTwitterScript()
+    }
```
Comment (macklinu) [inline packages/gatsby-plugin-twitter/src/gatsby-browser.js]: `injectedTwitterScript` always equals `false` - would that boolean need to be set to `true` the first time the `injectTwitterScript()` function is executed?

File: packages/gatsby/cache-dir/prefetch.js (hunk 2 above)
Comment (kkemple) [reply packages/gatsby/cache-dir/prefetch.js]: weird

File: www/src/components/search-form.js (hunk 4 above)
Comment (kkemple) [reply www/src/components/search-form.js]: nice

File: www/src/pages/showcase.js [hunk 6]
```diff
@@ -29,8 +29,8 @@ export const showcaseQuery = graphql`
           childScreenshot {
             screenshotFile {
               childImageSharp {
-                sizes(maxWidth: 512) {
-                  ...GatsbyImageSharpSizes
+                fluid(maxWidth: 512) {
```
Comment (kkemple) [inline www/src/pages/showcase.js]: would love to chat about the difference here at some point

Comment (KyleAMathews) [reply www/src/pages/showcase.js]: They're pretty much the same thing — we just renamed them as "sizes" and "resolutions" were near universally not understood :-D https://next.gatsbyjs.org/docs/migrating-from-v1-to-v2/#rename-responsive-image-queries

File: packages/gatsby-plugin-twitter/src/gatsby-browser.js (hunk 5 above)
Comment (KyleAMathews) [reply packages/gatsby-plugin-twitter/src/gatsby-browser.js]: haha yes, oops :-)

File: www/src/pages/showcase.js (hunk 6 above)
Comment (DSchau) [reply www/src/pages/showcase.js]: Totally unrelated, but this is such a good change :) It _always_ confused me which one was which.

Comment (KyleAMathews) [reply www/src/pages/showcase.js]: Glad to hear!

**Pull request merged**